packages/backend/src/app/workers/sandbox.ts

```diff
@@ -135,31 +168,79 @@ export class Sandbox {
 
 }
 
+
 export default class SandboxManager {
     private static _instance?: SandboxManager
 
-    private readonly queue: number[] = []
+    private readonly sandboxes: Map<number, Sandbox> = new Map()
+    private readonly mutex: Mutex = new Mutex()
 
     private constructor() {
         if (SandboxManager._instance != null) {
             throw new Error('Use Singleton.instance instead of new.')
         }
         for (let boxId = 0; boxId < 1000; ++boxId) {
-            this.queue.push(boxId)
+            this.sandboxes.set(boxId, new Sandbox({
+                boxId,
+                cached: false,
+                used: false,
+                resourceId: null,
+                lastUsed: 0,
+            }))
         }
         SandboxManager._instance = this
     }
 
-    obtainSandbox(): Sandbox {
-        const sandboxId = this.queue.pop()
-        if (sandboxId === undefined) {
-            throw new Error('Unexpected error, ran out of sandboxes')
+    async obtainSandbox(key: string): Promise<Sandbox> {
+        // Acquire the lock
+        const release = await this.mutex.acquire()
+
+        // Find sandbox with resourceId equal to key and not used
+        const sandbox = Array.from(this.sandboxes.values()).find(s => s.resourceId === key && !s.used)
+        if (sandbox) {
+            sandbox.used = true
+            sandbox.lastUsed = Date.now()
+            sandbox.cached = true
+
+            // Release the lock
+            release()
+
+            return sandbox
         }
-        return new Sandbox(sandboxId)
+
+        // Find oldest sandbox not used
+        const oldestSandbox = Array.from(this.sandboxes.values()).reduce((oldest, current) => {
+            if (oldest === null) {
+                return current
+            }
+            if (current.lastUsed < oldest.lastUsed) {
+                return current
+            }
+            return oldest
+        }, null)
```

Comment:
tip: if no initial value is provided for the accumelator, then the first element in the array is used and reduction starts from the second element in the array